contracts/src/teleporter/registry/TeleporterRegistryApp.sol

```diff
@@ -0,0 +1,268 @@
+// (c) 2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+// SPDX-License-Identifier: Ecosystem
+
+pragma solidity 0.8.25;
+
+import {TeleporterRegistry} from "./TeleporterRegistry.sol";
+import {ITeleporterReceiver} from "@teleporter/ITeleporterReceiver.sol";
+import {ITeleporterMessenger, TeleporterMessageInput} from "@teleporter/ITeleporterMessenger.sol";
+import {Context} from "@openzeppelin/contracts@5.0.2/utils/Context.sol";
+import {ReentrancyGuard} from "@openzeppelin/contracts@5.0.2/utils/ReentrancyGuard.sol";
+import {SafeERC20} from "@openzeppelin/contracts@5.0.2/token/ERC20/utils/SafeERC20.sol";
+import {IERC20} from "@openzeppelin/contracts@5.0.2/token/ERC20/IERC20.sol";
+
+/**
+ * @dev TeleporterRegistryApp provides upgrade utility for applications built on top
+ * of the Teleporter protocol by integrating with the {TeleporterRegistry}.
+ *
+ * This contract is intended to be inherited by other contracts that wish to use the
+ * upgrade mechanism. It provides an interface that restricts access to only Teleporter
+ * versions that are greater than or equal to `minTeleporterVersion`.
+ *
+ * @custom:security-contact https://github.com/ava-labs/teleporter/blob/main/SECURITY.md
+ */
+abstract contract TeleporterRegistryApp is Context, ITeleporterReceiver, ReentrancyGuard {
```

Comment:
Would it make sense to add a `ITeleporterRegistryApp` interface that both the upgradeable and non-upgradeable versions implement?

Reply:
To aid in review, I'd suggest we defer this to a separate PR.

Reply:
This may require more changes. Right now most of the functions are `public`, and for interfaces the visibility would need to be made `external`. The functions were made `public` so that inherited contracts can still call these functions internally without doing `this.func`. To still provide the ability to call internally, we could go through the external calling internal logic pattern.